Added basic covariance publishing based off the data sheet from BNO08…

Skel-Enkai opened the pull request:
I have added the basic variances to the Covariance matrices of the IMU message published by the bno08x. 
This will allow the package to interact more accurately with EKF filters, etc.

The source I used to find the variances can be found here:
 https://www.ceva-ip.com/wp-content/uploads/BNO080_085-Datasheet.pdf
 
 Please let me know if there are any stylistic changes I can make. 
 (The  set(CMAKE_EXPORT_COMPILE_COMMANDS ON) was just for my LSP server, I did not actually mean to commit it, but I'm too lazy to remove it x))

Skel-Enkai committed:
Added basic covariance publishing based off the data sheet from BNO08X, this is just the variances and not the covariances as these aren't supplied from the data sheet.

bnbhat commented on src/bno08x_ros.cpp:
```diff
@@ -242,6 +245,39 @@ void BNO08xROS::sensor_callback(void *cookie, sh2_SensorValue_t *sensor_value) {
+    // 3.5 degrees rotation vector error
+    // 3.1 degrees/s gyrometer error
+    // 0.35 m/s^2 linear acceleration error
+    double orientation_covariance = pow(degreesToRadians(3.5), 2);
```
I would put this calculation somewhere in the init functions or in the constructor so that it only run once.

Skel-Enkai replied:
Should I set them as parameters so they are accessible by the configuration file?

bnbhat replied:
Since it is a static value, even if you hardcode it, it should be okay. But parameter would be great "in case" if some one want to change it :)

bnbhat commented on include/bno08x_driver/geom.hpp:
I would rather not create a separate file for it. you can even skip the fn. I would just put `degree * M_PI / 180.0;` in line.

Skel-Enkai replied:
Will do inline.

bnbhat commented on CMakeLists.txt:
```diff
@@ -14,6 +14,7 @@ if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
 endif()
 
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-old-style-declaration")
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
```
if you remove the geom.hpp/cpp, you'll not need this

bnbhat commented on src/geom.cpp:
You could also remove this

bnbhat reviewed (changes requested):
Thank you for your PR. I appreciate it.

I have left some comments. Also could you please point me where the covariences/accuracy is mentioned in the docs?, if you already know where it is :)

Skel-Enkai commented:
The performance characteristics are found on page 49/59, under section: 6.7 Performance Characteristics
I will fix these issues sometime tomorrow, and resubmit my pull request, thank you for the fast response!

bnbhat commented:
> The performance characteristics are found on page 49/59, under section: 6.7 Performance Characteristics I will fix these issues sometime tomorrow, and resubmit my pull request, thank you for the fast response!

Thank you, I'll take a look at the datasheet. 

Take your time! resubmit it when you can :)

Skel-Enkai commented:
I couldn't help myself but do it now, I believe I have fixed all issues raised. Let me know if there's anything else I can do. 

I will be in the lab next week, and will test that these covariance values return reasonable results with an ekf filter like robot_localisation.

Skel-Enkai committed:
Fixed various style issues and added parameters for the added variances.
Removed one function source code file, in favour of inline converions.
Placed time stamping messages just before message publication, to ensure
fidelity and accuracy of stamping.

Skel-Enkai committed:
After inspection, it's clear that the same IMU message in memeory is reused. Rather than redunantly setting the covariances every time beore publish, I have set an init function.

bnbhat reviewed (approved):
LGTM

Skel-Enkai committed:
Created config parameter for full orientation covariance matrix.

Richard-poiqwer committed:
Updated bno08x_ros.cpp for IMU Covariance Matrices

Updated code to copy pattern for orientation covariance for both gyrometer and linear covariance

Richard-poiqwer committed:
Altered bno08x_ros.hpp to include covariance matrices

Added the gyrometer and linear covaraince including their default matrices

Richard-poiqwer committed:
Updated bno085_i2c.yaml to include additional covariance

Added the gyrometer and linear covariance matrices

Richard-poiqwer committed:
Update bno085_i2c.yaml

Changed default values of gyrometer and linear covariance

Richard-poiqwer committed:
Update bno08x_ros.cpp

Indentation between covariance declarations

Skel-Enkai committed:
Merge pull request #2 from Skel-Enkai/IMU_Covariance_Matrix_Changes

Added ROS2 parameter configuration for all sensor covariance matrices.

Skel-Enkai committed:
Fixed incorrect field name access for sensor_msgs:IMU.
